core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -576,96 +583,252 @@ public void addPhysicallyWrittenRows(long rows) {
         metrics.tableWriter().addPhysicallyWrittenRows(rows);
     }
 
-    public int attachPartition(long timestamp) {
+    public StatusCode attachPartition(long timestamp) {
         return attachPartition(timestamp, true);
     }
 
-    public int attachPartition(long timestamp, boolean validateDataFiles) {
+    public StatusCode attachPartition(long timestamp, boolean validateDataFiles) {
         // Partitioned table must have a timestamp
-        // SQL compiler will check that table is partitioned
+        // SQL compiler will check that table has it
         assert metadata.getTimestampIndex() > -1;
 
-        CharSequence timestampCol = metadata.getColumnQuick(metadata.getTimestampIndex()).getName();
         if (txWriter.attachedPartitionsContains(timestamp)) {
-            LOG.info().$("partition is already attached [path=").$(path).$(']').$();
-            return PARTITION_ALREADY_ATTACHED;
+            LOG.info().$("partition is already attached [path=").$(path).I$();
+            // TODO: potentially we can merge with existing data
+            return StatusCode.PARTITION_ALREADY_ATTACHED;
         }
 
         boolean rollbackRename = false;
         try {
-            setPathForPartition(path, partitionBy, timestamp, false);
-            if (!ff.exists(path.$())) {
-                setPathForPartition(other, partitionBy, timestamp, false);
-                other.put(DETACHED_DIR_MARKER);
-
-                if (ff.exists(other.$())) {
-                    if (ff.rename(other, path) == Files.FILES_RENAME_OK) {
-                        rollbackRename = true;
-                        LOG.info().$("moved partition dir: ").$(other).$(" to ").$(path).$();
-                    } else {
-                        throw CairoException.instance(ff.errno()).put("File system error on trying to rename [from=")
-                                .put(other).put(",to=").put(path).put(']');
-                    }
-                }
+            if (detachedPath == null) {
+                other2 = new Path();
+                detachedPath = new Path().of(configuration.getDetachedRoot()).concat(tableName);
+                detachedRootLen = detachedPath.length();
             }
 
-            if (ff.exists(path.$())) {
-                // find out lo, hi ranges of partition attached as well as size
-                final long partitionSize = readPartitionSizeMinMax(ff, path, timestampCol, tempMem16b, timestamp);
-                if (partitionSize > 0) {
+            // final name of partition folder after attach
+            setPathForPartition(path.trimTo(rootLen), partitionBy, timestamp, false);
```

Comment:
As discussed it should be

```
TableUtils.txnPartitionConditionally(path, getTxn());
```

Reply:
done in [47a14c2](https://github.com/questdb/questdb/pull/2280/commits/47a14c2e53d813b5612f72cf047cde353adfd205)